flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkBase.java

```diff
@@ -0,0 +1,234 @@
+
+		if (params.has(CONFIG_KEY_BULK_FLUSH_MAX_ACTIONS)) {
+			bulkProcessorFlushMaxActions = params.getInt(CONFIG_KEY_BULK_FLUSH_MAX_ACTIONS);
+			userConfig.remove(CONFIG_KEY_BULK_FLUSH_MAX_ACTIONS);
+		} else {
+			bulkProcessorFlushMaxActions = null;
+		}
+
+		if (params.has(CONFIG_KEY_BULK_FLUSH_MAX_SIZE_MB)) {
+			bulkProcessorFlushMaxSizeMb = params.getInt(CONFIG_KEY_BULK_FLUSH_MAX_SIZE_MB);
+			userConfig.remove(CONFIG_KEY_BULK_FLUSH_MAX_SIZE_MB);
+		} else {
+			bulkProcessorFlushMaxSizeMb = null;
+		}
+
+		if (params.has(CONFIG_KEY_BULK_FLUSH_INTERVAL_MS)) {
+			bulkProcessorFlushIntervalMillis = params.getInt(CONFIG_KEY_BULK_FLUSH_INTERVAL_MS);
+			userConfig.remove(CONFIG_KEY_BULK_FLUSH_INTERVAL_MS);
+		} else {
+			bulkProcessorFlushIntervalMillis = null;
+		}
+
+		this.userConfig = userConfig;
+	}
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
+		client = clientFactory.create(userConfig);
+
+		BulkProcessor.Builder bulkProcessorBuilder = BulkProcessor.builder(
+			client,
+			new BulkProcessor.Listener() {
+				@Override
+				public void beforeBulk(long executionId, BulkRequest request) { }
+
+				@Override
+				public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+					if (response.hasFailures()) {
+						for (BulkItemResponse itemResp : response.getItems()) {
+							if (itemResp.isFailed()) {
+								LOG.error("Failed to index document in Elasticsearch: " + itemResp.getFailureMessage());
+								failureThrowable.compareAndSet(null, new RuntimeException(itemResp.getFailureMessage()));
+							}
+						}
+						hasFailure.set(true);
+					}
+				}
+
+				@Override
+				public void afterBulk(long executionId, BulkRequest request, Throwable failure) {
+					LOG.error(failure.getMessage());
+					failureThrowable.compareAndSet(null, failure);
+					hasFailure.set(true);
+				}
+			}
+		);
+
+		// This makes flush() blocking
+		bulkProcessorBuilder.setConcurrentRequests(0);
+
+		if (bulkProcessorFlushMaxActions != null) {
+			bulkProcessorBuilder.setBulkActions(bulkProcessorFlushMaxActions);
+		}
+
+		if (bulkProcessorFlushMaxSizeMb != null) {
+			bulkProcessorBuilder.setBulkSize(new ByteSizeValue(bulkProcessorFlushMaxSizeMb, ByteSizeUnit.MB));
+		}
+
+		if (bulkProcessorFlushIntervalMillis != null) {
+			bulkProcessorBuilder.setFlushInterval(TimeValue.timeValueMillis(bulkProcessorFlushIntervalMillis));
+		}
+
+		bulkProcessor = bulkProcessorBuilder.build();
+		requestIndexer = new BulkProcessorIndexer(bulkProcessor);
+	}
+
+	@Override
+	public void invoke(T value) throws Exception {
+		elasticsearchSinkFunction.process(value, getRuntimeContext(), requestIndexer);
+	}
+
+	@Override
+	public void close() throws Exception {
+		if (bulkProcessor != null) {
+			bulkProcessor.close();
+			bulkProcessor = null;
+		}
+
+		if (client != null) {
+			client.close();
+			client = null;
+		}
+
+		clientFactory.cleanup();
+
+		if (hasFailure.get()) {
```

Comment:
I would suggest to first merge this PR and then work on FLINK-5353 to not mix up too many things at the same time.